File: Codes/Level 2/Two Pointers/TwoSumDesign.cpp
class TwoSum {
public:
    /**
     * @param number: An integer
     * @return: nothing
     */
    vector<int> v;
    void add(int number) {
        // write your code here
        v.push_back(number);
    }

    /**
     * @param value: An integer
     * @return: Find if there exists any pair of numbers which sum is equal to the value.
     */
    bool find(int value) {
        // write your code here
        sort(v.begin(), v.end());
        int start = 0, end = v.size() - 1;
        while(start < end){
            int sum = v[start] + v[end];
            if(sum < value)
                start++;
            else if(sum > value)
                end--;
            else
                return true;
        }
        return false;
    }
};

File: Codes/Level 2/Two Pointers/TwoSumProblems/TwoSumDesign.java
public class TwoSum {
    /**
     * @param number: An integer
     * @return: nothing
     */

    ArrayList<Integer> stream;

    public void add(int number) {
        // write your code here
        if(stream == null)  stream = new ArrayList<>();
        stream.add(number);
    }

    /**
     * @param value: An integer
     * @return: Find if there exists any pair of numbers which sum is equal to the value.
     */
    public boolean find(int value) {
        // write your code here
        Collections.sort(stream);
        int left = 0, right = stream.size() - 1;
        while(left < right){
            int currSum = stream.get(left) + stream.get(right);
            if(currSum == value)
                return true;
            else if(currSum < value)
                left++;
            else
                right--;
        }
        return false;
    }
}